runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java

```diff
@@ -104,6 +105,9 @@ public void evaluate(Flatten.FlattenPCollectionList<T> transform, EvaluationCont
           }
           unionRDD = context.getSparkContext().union(rdds);
         }
+        if (cacheHint) {
```

Comment:
This comment is for all following:
```java
if (cacheHint) {
  rdd.cache();
}
```
and
```java
if (cacheHint) {
  dstream.cache();
}
``` 
I'd go for using the runner's `Dataset` so that we use the user-defined `StorageLevel` (batch). This will require slight changes, and also get rid of `multiReads` optimization in `EvaluationContext` since this is a better cache optimization.

I'd also try something more fluent like:

```java
cacheHint ? rdd.cache : rdd;
``` 
Though you'll probably hide it in Dataset anyway so I'm not sure it will do much difference.